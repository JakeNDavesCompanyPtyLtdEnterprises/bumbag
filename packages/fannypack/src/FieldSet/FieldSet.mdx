---
name: FieldSet
route: /form/FieldSet
menu: Form
---

import { Playground } from 'docz';
import PropsTable from '../_docs/components/PropsTable';
import { Box } from '../primitives/index';
import { Button, InputField } from '../index';
import FieldSet, { FieldSet as FieldSetProps } from './FieldSet';

# FieldSet

## Import

`import { FieldSet } from 'fannypack'`

## Basic Usage

The `<FieldSet>` component is a utility belt that wraps around field elements to space them out correctly. It uses the [`<Set>` utility](/utils/set) internally.

<Playground>
  <FieldSet>
    <FieldSet isHorizontal>
      <InputField name="firstName" label="First name" />
      <InputField name="lastName" label="Last name" />
    </FieldSet>
    <InputField name="username" label="Username" />
    <InputField name="password" label="Password" type="password" />
    <Button>Save</Button>
  </FieldSet>
</Playground>

## Props

<PropsTable component={FieldSetProps} />

## Theming

### Schema

```jsx
{
  FieldSet: {
    base: css | Object,
    vertical: css | Object,
    horizontal: css | Object
  }
}
```
