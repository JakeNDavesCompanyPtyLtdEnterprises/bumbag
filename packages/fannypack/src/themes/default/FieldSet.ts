// @ts-ignore
import _get from 'lodash/get';

export default (overrides: any) => ({
  spacing: 'major-3',
  ..._get(overrides, 'FieldSet', {})
});
