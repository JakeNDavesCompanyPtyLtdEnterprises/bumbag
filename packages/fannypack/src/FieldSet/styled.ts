import { theme } from 'styled-tools';

import styled, { css } from '../styled';
import Set from '../Set';
import { FieldSetProps } from './FieldSet';

export default styled(Set)<FieldSetProps>`
  ${props =>
    props.isHorizontal
      ? css`
          display: flex;

          & > * {
            flex: 1;
          }

          @media screen and (max-width: ${theme('fannypack.layout.tabletBreakpoint')}px) {
            display: block;
            margin-top: 0rem;
          }

          & {
            ${theme('fannypack.FieldSet.horizontal')};
          }
        `
      : css`
          display: block;

          & {
            ${theme('fannypack.FieldSet.vertical')};
          }
        `};

  & {
    ${theme('fannypack.FieldSet.base')};
  }
`;
