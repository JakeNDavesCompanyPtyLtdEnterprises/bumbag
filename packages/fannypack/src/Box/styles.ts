import _get from 'lodash/get';
import { css, cssClass } from '../styled';
import { breakpoint, theme } from '../utils';
import { ThemeConfig } from '../types';

export const style = styleProps => cssClass`
  & {
    ${styleProps.style};
    ${buildVisibleAttributes(styleProps)};
  }

  ${styleProps.altitude &&
    css`
      & {
        ${theme(`altitudes.${styleProps.altitude}`)(styleProps)};
      }
    `};
`;

export const Box = styleProps => cssClass`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme(`${styleProps.themeKey}.css.root`)(styleProps)};
  }
`;

function buildVisibleAttributes(props: { hiddenBreakpoint?: string; showBreakpoint?: string; theme: ThemeConfig }) {
  let _breakpoint = props.hiddenBreakpoint || props.showBreakpoint;
  if (!_breakpoint) return;

  return breakpoint(
    _breakpoint,
    css`
      display: none;
    `,
    { show: Boolean(props.showBreakpoint) }
  )(props);
}
