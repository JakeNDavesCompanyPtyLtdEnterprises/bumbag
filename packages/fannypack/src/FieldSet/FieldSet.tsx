/* eslint-disable react/prop-types */
import * as React from 'react';
import * as PropTypes from 'prop-types';
// @ts-ignore
import _get from 'lodash/get';

import { withTheme } from '../styled';
import { SetProps } from '../Set/Set';
import _FieldSet from './styled';

export type LocalFieldSetProps = {
  children: React.ReactNode;
  isHorizontal?: boolean;
  spacing?: string;
  theme?: Object;
};
export type FieldSetProps = SetProps & LocalFieldSetProps;

export const FieldSet: React.FunctionComponent<LocalFieldSetProps> = ({ children, theme, ...props }) => {
  const spacing = props.spacing || _get(theme, 'fannypack.FieldSet.spacing');
  return (
    <_FieldSet isVertical={!props.isHorizontal} {...props} spacing={spacing}>
      {children}
    </_FieldSet>
  );
};

FieldSet.propTypes = {
  children: PropTypes.node.isRequired,
  isHorizontal: PropTypes.bool,
  spacing: PropTypes.string
};

FieldSet.defaultProps = {
  isHorizontal: false,
  spacing: undefined
};

// @ts-ignore
const C: React.FunctionComponent<FieldSetProps> = withTheme(FieldSet);
export default C;
