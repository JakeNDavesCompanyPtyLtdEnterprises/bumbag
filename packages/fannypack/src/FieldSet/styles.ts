import { css, cssClass } from '../styled';
import { breakpoint, theme } from '../utils';

export const FieldSet = styleProps => cssClass`
${
  styleProps.isHorizontal
    ? css`
        display: flex;

        & > * {
          flex: 1;
        }

        ${breakpoint(
          `max-tablet`,
          css`
            display: block;
            margin-top: 0rem;
          `
        )(styleProps)}

        & {
          ${theme(`${styleProps.themeKey}.css.horizontal`)(styleProps)};
        }
      `
    : css`
        display: block;

        & {
          ${theme(`${styleProps.themeKey}.css.vertical`)(styleProps)};
        }
      `
};

  & {
    ${theme(`${styleProps.themeKey}.css.root`)(styleProps)};
  }
`;
