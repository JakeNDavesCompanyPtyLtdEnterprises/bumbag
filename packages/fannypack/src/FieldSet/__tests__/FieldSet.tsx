import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Box } from '../../Box';
import { FieldSet } from '../FieldSet';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(
      <FieldSet>
        <Box />
        <Box />
      </FieldSet>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <FieldSet color="primary">
        <Box />
        <Box />
      </FieldSet>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with spacing prop', () => {
    const { container } = render(
      <FieldSet spacing="major-4">
        <Box />
        <Box />
      </FieldSet>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with isHorizontal prop', () => {
    const { container } = render(
      <FieldSet isHorizontal>
        <Box />
        <Box />
      </FieldSet>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(
        <FieldSet use="div">
          <Box />
          <Box />
        </FieldSet>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with FieldSet props', () => {
      const { result } = renderHook(() => FieldSet.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(
        <FieldSet>
          {FieldSetProps => (
            <div {...FieldSetProps}>
              <Box />
              <Box />
            </div>
          )}
        </FieldSet>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('FieldSet.root should render correctly', () => {
    const { container } = render(
      <FieldSet>
        <Box />
        <Box />
      </FieldSet>,
      {
        // @ts-ignore
        theme: { FieldSet: { css: { root: { backgroundColor: 'red' } } } }
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(
      <FieldSet>
        <Box />
        <Box />
      </FieldSet>,
      {
        // @ts-ignore
        theme: { FieldSet: { defaultProps: { className: 'test', color: 'primary' } } }
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
