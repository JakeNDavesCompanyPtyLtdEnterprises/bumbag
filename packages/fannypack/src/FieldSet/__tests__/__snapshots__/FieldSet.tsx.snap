// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`composition hook should return with FieldSet props 1`] = `
Object {
  "className": "css-1tsdp7n css-w4rn5t css-1pwrqjj css-3ekj6a fp-Box fp-Set fp-FieldSet",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

.emotion-7 {
  color: #3722ed !important;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-7 test fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

.emotion-7 {
  color: #3722ed !important;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-7 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with isHorizontal prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -1.5rem;
  margin-top: -1.5rem;
}

.emotion-5 > * {
  margin-left: 1.5rem;
  margin-top: 1.5rem;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 > * {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media screen and (max-width:768px) {
  .emotion-4 {
    display: block;
    margin-top: 0rem;
  }
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`props should render correctly with spacing prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  display: block;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 2rem;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;

exports[`theming FieldSet.root should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-5 > *:not(:last-child) {
  margin-bottom: 1.5rem;
}

.emotion-4 {
  display: block;
}

.emotion-4 {
  background-color: red;
}

<div
  class="emotion-4 emotion-5 emotion-0 emotion-1 fp-Box fp-Set fp-FieldSet"
>
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
  <div
    class="emotion-0 emotion-1 fp-Box"
  />
</div>
`;
